File: Java_Demos/MountainBike.java
public class MountainBike extends Bicycle {
        
    // the MountainBike subclass adds one field
    public int seatHeight;

    // the MountainBike subclass has one constructor
    public MountainBike(int startHeight,
                        int startCadence,
                        int startSpeed,
                        int startGear) {
        super(startCadence, startSpeed, startGear);
        seatHeight = startHeight;
    }   
        
    // the MountainBike subclass adds one method
    public void setHeight(int newValue) {
        seatHeight = newValue;
    }

    // Method to print bike attributes
    public void PrintBikeAttributes(){
        System.out.println("Height: " + this.seatHeight);
        System.out.println("Cadence: " + this.cadence);
        System.out.println("Speed: " + this.speed);
        System.out.println("Gear: " + this.gear);
    }

    public static void main(String args[]) {
        MountainBike mBike = new MountainBike(2, 10, 2, 1);
        System.out.println("Initial bike attributes");
        mBike.PrintBikeAttributes();
        mBike.setCadence(30);
        mBike.setGear(2);
        mBike.speedUp(10);
        System.out.println("Bike attributes after gearing up");
        mBike.PrintBikeAttributes();
        
    }   
}

class Bicycle {
        
    // the Bicycle class has three fields
    public int cadence;
    public int gear;
    public int speed;
        
    // the Bicycle class has one constructor
    public Bicycle(int startCadence, int startSpeed, int startGear) {
        gear = startGear;
        cadence = startCadence;
        speed = startSpeed;
    }
        
    // the Bicycle class has four methods
    public void setCadence(int newValue) {
        cadence = newValue;
    }
        
    public void setGear(int newValue) {
        gear = newValue;
    }
        
    public void applyBrake(int decrement) {
        speed -= decrement;
    }
        
    public void speedUp(int increment) {
        speed += increment;
    }
        
}

File: Kotlin/inheritance_demo2/MountainBike.kt
open class Bicycle {
    // the Bicycle class has three fields
    private var cadence: Int
    private var gear: Int
    private var speed: Int

    // the Bicycle class has one constructor
    constructor(startCadence: Int, startSpeed: Int, startGear: Int) {
        this.gear = startGear
        this.cadence = startCadence
        this.speed = startSpeed
    }

    // the Bicycle class has four methods
    fun setCadence(newValue: Int) {
        this.cadence = newValue
    }

    fun setGear(newValue: Int) {
        this.gear = newValue
    }

    fun applyBrake(decrement: Int) {
        this.speed -= decrement
    }

    fun speedUp(increment: Int) {
        this.speed += increment
    }

    fun printBicyleAttributes() {
        println("Cadence: " + this.cadence);
        println("Speed: " + this.speed);
        println("Gear: " + this.gear);
    }
}

class MountainBike : Bicycle {
    // the MountainBike subclass adds one field
    private var seatHeight: Int;

    // the MountainBike subclass has one constructor
    constructor(startHeight: Int,
                startCadence: Int,
                startSpeed: Int,
                startGear: Int) : super(startCadence, startSpeed, startGear) {
        this.seatHeight = startHeight;
    }

    // the MountainBike subclass adds one method
    fun setHeight(newValue: Int) {
        this.seatHeight = newValue;
    }

    // Method to print bike attributes
    fun PrintBikeAttributes() {
        println("Height: " + this.seatHeight);
        this.printBicyleAttributes()
    }
}

fun main(args: Array<String>)
{
    val mBike = MountainBike(2, 10, 2, 1);
    println("Initial bike attributes");
    mBike.PrintBikeAttributes();
    mBike.setCadence(30);
    mBike.setGear(2);
    mBike.speedUp(10);
    println("Bike attributes after gearing up");
    mBike.PrintBikeAttributes();

}